---
title: "June 13 Code"
author: "Leila Curtis"
date: "2024-06-13"
output: html_document
---

```{r Set Up, eval=F}
srun --constraint=ib -p lotterhos --pty /bin/bash #Run on the lotterhos partition
module load singularity/3.10.3 #Load singularity
export PATH=/home/curtis.lei/lumpy-sv/bin/samtools-1.20:$PATH #Load samtools

cd lumpy-sv #Change directory to lumpy directory
```

```{r Data Alignment, eval=F}
cp /work/lotterhos/2020_CodGenomes/labeled_bam_Out/Pop1_16216.f.rg.bam /home/curtis.lei/lumpy-sv #Copy the .bam file to my directory

# Extract the discordant paired-end alignments.
samtools view -b -F 1294 Pop1_16216.f.rg.bam > Pop1_16216.discordants.unsorted.bam

# Extract the split-read alignments
samtools view -h Pop1_16216.f.rg.bam \
    | scripts/extractSplitReads_BwaMem -i stdin \
    | samtools view -Sb - \
    > Pop1_16216.splitters.unsorted.bam

# Sort both alignments 
samtools sort Pop1_16216.discordants.unsorted.bam -o Pop1_16216.discordants.bam
samtools sort Pop1_16216.splitters.unsorted.bam -o Pop1_16216.splitters.bam
```

```{r LUMPY traditional, eval=F}
#Generate empirical insert size statistics on library
samtools view Pop1_16216.f.rg.bam \
    | tail -n+100000 \
    | scripts/pairend_distro.py \
    -r 101 \
    -X 4 \
    -N 10000 \
    -o Pop1_16216.lib1.histo

#Run with paired-end and split reads
singularity run /work/lotterhos/Containers/lumpy_0.2.13.sif \
    -mw 4 \
    -tt 0 \
    -pe id:Pop1_16216,bam_file:Pop1_16216.discordants.bam,histo_file:Pop1_16216.lib1.histo,mean:158.892146073,stdev:71.7231558068,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
    -sr id:Pop1_16216,bam_file:Pop1_16216.splitters.bam,back_distance:10,weight:1,min_mapping_threshold:20 \
    > Pop1_16216.vcf
```
